### ISSUE TITLE
Quests API suggestion!

### ISSUE THREAD

**Issue by LuckyClutch07** [FlyingPikachu/Quests#564]
Please please please implement this I beg you, Improve the API please don't make devs write JARS for modules for simple things like a boolean ;-; please please please, make a registerRequirement() method i beg i assure you it isn't too hard I mean is probably harder to load the jars than to load the class anyway that is what you currently do you loop thru the jar and search for the "main" class and load it in so why not just make a method to load the things in directly without having to put the jar in the modules ? now if there is already a method for this I do really really apologize for my ignorance but I haven't found please look into this 👍

**Comment by FlyingPikachu** [comment]
Not exactly sure what it is you’re asking. You want Quests to load non-jar files, like scripts? This is already possible through Denizen iirc.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by LuckyClutch07** [comment]
no, not with script ok so i have Quests as a dependency on my plugin and i hooked into it and i am loading before Quests etc... so i want you to add a method to load requirements into Quests etc.

**Comment by FlyingPikachu** [comment]
Due to the way Quests manages it's files, all requirements, custom or otherwise, must be ready at load time (when the server starts). This is partially why Quests uses modules, but it's mostly to give the server owner some control over whether they _want_ a requirement or not. Dropping a module in the folder is a form of cheap authentication.

If I were to allow plugins to "inject" a requirement (or reward or objective), there would be no way for a server owner to refuse it. Granted they wouldn't have to _use_ it, but let's say the requirement doesn't work right or includes a bunch of slow or undesirable background tasks _that run regardless of its use_ (after all, Quests can't assume all devs would code responsibly). In order to stop this, the owner can't just go and remove a module - they would have to remove the offender's entire plugin (or disable some config option somewhere in that plugin, _assuming_ it even has one and it works).

Tl;dr current method allows Quests to maintain some degree of control. Not seeing any advantage to class registration beyond ease of use.

**Comment by frost-byte** [comment]
While I can see your reasoning - there are still plenty of cases where a user of Quests would have to do additional configuration to use a binding to other plugins either way.  Even as a server owner that does not do any coding, you still have to realistically deal with configuration of plugins, it's practically unavoidable.

One of the things I do not like about the current approach - I have to create a stand alone intermediary between both Quests and my Plugin, and then I have to provide a way to version that separately from my plugin.  Then I have to include it in my build process somehow.  One of the many reasons we use maven or gradle is to greatly simplify the management of libraries (jars) and their versions, it seems strange to not leverage them.

In order for my plugin to work with your plugin, I have to create an entirely separate jar that isn't a library or second plugin, which does not seem like standard practice.

As a plugin creator I would have to create an API for my plugin that provides all of the functionality needed within an intermediary jar and then connect that to the current api that you've created.

If you would like to see how some other plugin creators have created an API the allows for an extendable or configurable system I can provide you a few examples:

- Player Analytics - [PluginData API](https://github.com/Rsl1122/Plan-PlayerAnalytics/wiki/APIv4)
- [Annotation Command Framework](https://github.com/aikar/commands/wiki/Using-ACF) - System for registering all of your plugin commands; includes completions, contexts, replacements, conditions and Locales

I think Player Analytics provides functionality that is more in tune with what you might do.
He provides many bindings for between other plugins and the Player Analytics system - each provides custom data from those plugins and makes them presentable to Player Analytics.  The config file for Plan also allows you to designate which plugins it can use.

My use case is a little different than most, I'm not creating a standalone, drop in plugin that's going to be widely used, and as far as I can tell, my needs are currently met.  Still, I thought I'd offer you some examples and reasoning as to why you might want to try another approach.

One thing you might do is create a skeleton/blue print project that others can use as a starting point to create their own Custom jars.
